src/class/audio/audio_device.h

```diff
@@ -493,68 +458,83 @@ TU_ATTR_WEAK bool tud_audio_rx_done_post_read_cb(uint8_t rhport, uint16_t n_byte
+// Note that due to a bug in its USB Audio 2.0 driver, Windows currently requires 16.16 format for _all_ USB 2.0 devices. On Linux and macOS it seems the 
 // driver can work with either format. So a good compromise is to keep format correction disabled and stick to 16.16 format.
 
 // Feedback value can be determined from within the SOF ISR of the audio driver. This should reduce jitter. If the feature is used, the user can not set the feedback value.
 
+// Determine feedback value - The feedback method is described in 5.12.4.2 of the USB 2.0 spec
+// Boiled down, the feedback value Ff = n_samples / (micro)frame.
+// Since an accuracy of less than 1 Sample / second is desired, at least n_frames = ceil(2^K * f_s / f_m) frames need to be measured, where K = 10 for full speed and K = 13 for high speed, f_s is the sampling frequency e.g. 48 kHz and f_m is the cpu clock frequency e.g. 100 MHz (or any other master clock whose clock count is available and locked to f_s)
+// The update interval in the (4.10.2.1) Feedback Endpoint Descriptor must be less or equal to 2^(K - P), where P = min( ceil(log2(f_m / f_s)), K)
+// feedback = n_cycles / n_frames * f_s / f_m in 16.16 format, where n_cycles are the number of main clock cycles within fb_n_frames
+
 bool tud_audio_n_fb_set(uint8_t func_id, uint32_t feedback);
 static inline bool tud_audio_fb_set(uint32_t feedback);
 
 uint8_t tud_audio_n_get_fb_n_frames(uint8_t func_id);
 static inline uint8_t tud_audio_get_fb_n_frames();
 
-TU_ATTR_WEAK void tud_audio_sof_isr_cb(uint8_t func_id, uint32_t frame);
+#endif // CFG_TUD_AUDIO_ENABLE_EP_OUT && CFG_TUD_AUDIO_ENABLE_FEEDBACK_EP
 
-#endif // (CFG_TUD_AUDIO_ENABLE_FEEDBACK_DETERMINATION_OPTION == CFG_TUD_AUDIO_ENABLE_FEEDBACK_DETERMINATION_OPTION_SOF_BY_USER)
+// Update feedback value with passed cycles since last time this update function is called.
+// Typically called within tud_audio_sof_isr(). Required tud_audio_feedback_params_cb() is implemented
+// This function will also call tud_audio_feedback_set()
+// return feedback value in 16.16 for reference (0 for error)
+uint32_t tud_audio_feedback_update(uint8_t func_id, uint32_t cycles);
+
+enum {
+  AUDIO_FEEDBACK_METHOD_DISABLED,
+  AUDIO_FEEDBACK_METHOD_FREQUENCY_FIXED,
+  AUDIO_FEEDBACK_METHOD_FREQUENCY_FLOAT,
+  AUDIO_FEEDBACK_METHOD_FREQUENCY_POWER_OF_2,
+  AUDIO_FEEDBACK_METHOD_FIFO_COUNT_FIXED,
+  AUDIO_FEEDBACK_METHOD_FIFO_COUNT_FLOAT
+};
+
+typedef struct {
+  uint8_t method;
+  union {
+    struct {
+        uint32_t sample_freq;
+        uint32_t mclk_freq;
+    }frequency;
+
+    struct {
+        uint32_t nominal;
+        uint32_t threshold;
+        // variation etc ..
+    }fifo_count;
+  };
+}audio_feedback_params_t;
+
+
+// TU_ATTR_WEAK void tud_audio_feedback_params_cb(uint8_t func_id, uint8_t alt_itf, audio_feedback_params_t* feedback_param);
```

Comment:
@HiFiPhile I made a draft idea on how to add the fifo count method, the params_cb with struct for input. Fifo count method seems to reply on a few params: std (nominal) value, variation, and buffer threshold. The buffer threshold is a bit tricky, since it requires application to respect this. The rest can be computed within the stack. Let me know if I miss something before adding the actual code.

Reply:
I like the idea of union.

> The buffer threshold is a bit tricky, since it requires application to respect this.
Yes indeed. It can be calculated in the stack because normally it's N*PacketSize, or add a macro for easier threshold calculation in application.

https://github.com/HiFiPhile/i2s_bridge/blob/1d6a24fdd0f70d150a88d85d63ed4b583b3b6b30/audio_callbacks.c#L186
Here 8000 is for HS (125us), FS need 1000.

Reply:
@HiFiPhile I have a silly question, `fifo_count_flt` stands for float in 16.16 format right ? 

https://github.com/HiFiPhile/i2s_bridge/blob/1d6a24fdd0f70d150a88d85d63ed4b583b3b6b30/audio_callbacks.c#L113

Reply:
It's "filtered" fifo count. As fifo count is jumpy each time a packet is arrived / consumed, it's hard to regulate the fifo level based on raw count. So I make an average of fifo count to get a stabilized value.
Basically fifo_count_flt = `fifo_count_old * 255 + fifo_count`, I shifted the whole value <<8 to let the result has enough precision so float is not required.

Reply:
thanks @HiFiPhile , I kind of getting the idea, filtered ~ average here

```
filtered = 255 * old + raw_count
feedback = nominal - (filtered/256 - threshold)
```

out of curiosity, how would you come up with the filtered/avg formula and how does it depend on sample rate, FS/HS link. Sorry for more question, I am trying to figure out the big picture and normalize the argument for this compute method.

Reply:
> out of curiosity, how would you come up with the filtered/avg formula and how does it depend on sample rate, FS/HS link. 

I used J-Scope to watch FIFO level, it's kind of trial and error :)

![image](https://user-images.githubusercontent.com/4375114/171142831-75bed357-1ddd-4122-a552-6e2e39278dbe.png)

FIFO is filled by USB packets and drained by I2S.

For the **same** sample rate, since HS packets arrive 8x faster every packet is 8x smaller, FIFO level variation is much smaller so easier to filter.

The higher the sample rate the higher the variation, in theory a higher filter coefficient is needed to achieve a same smoothed result.
But since `feedback_std` is also proportional to sample rate, which cancelled the need of a higher filter coefficient.
So no need to change filter coefficient on sample rate change.

Reply:
thanks for the explanation, unfortunately, I still pretty much have no ideas how to generalize this into generic code. Apparently, there is no formula to fit all, I am just hoping to make some skeleton and allow user to simply tweaking a parameter or two with actual hardware. However, this is really out of my knowledge, and don't actually have any real application on hardware to test these out. 

Since this has been pending for quite some time, I would like to merge this as it is and come back later to this later on. The good news is we have an decent API with place holders union for fifo count method (or any new method we could find afterwards). Should you have time feel free to make an PR to enhance this :)